util.cc

```diff
@@ -1233,4 +1233,70 @@ string open_temp_fstream(ofstream &f, const string &name_template, const string
     return string(name.data());
 }
 
+/** @name Segmented read and write functions. */
+//@{
+
+// The maximum number of bytes that can be safely written in a single call
+// to ostream::write() is typically limited by std::streamsize, which is
+// often a signed 32-bit integer (2^31) - 1.
+constexpr int64_t MAX_SEGMENT_SIZE = (static_cast<int64_t>(1) << 31) - 1;
+
+/**
+ * @brief Ensure that no more than (2^31)-1 bytes are ever written by ostream::write()
+ * @param out The ostream& where the bytes are to be written
+ * @param val The bytes to write
+ * @param num_bytes The total number of bytes to write
+ */
+void segmented_write(ostream &out, const char *val, int64_t num_bytes) {
+    int64_t bytes_remaining = num_bytes;
+    const char *current_ptr = val; // Use a mutable pointer to advance through the buffer
+
+    while (bytes_remaining > 0) {
+        // Determine the size of the current chunk to write
+        const auto current_chunk_size = static_cast<std::streamsize>(std::min(bytes_remaining, MAX_SEGMENT_SIZE));
+
+        // Call ostream::write() with the current segment
+        out.write(current_ptr, current_chunk_size);
+        if (out.fail()) {
+            throw InternalErr(__FILE__, __LINE__, "Error writing to a stream");
+        }
+
+        current_ptr += current_chunk_size;
+
+        bytes_remaining -= current_chunk_size;
+    }
+}
+
+/**
+ * @brief Reads a large number of bytes from an istream in segments, matching segmented_write.
+ * @param in The istream& from where the bytes are to be read.
+ * @param buffer The char* buffer where the bytes will be read into.
+ * @param num_bytes_to_read The total number of bytes to read.
+ */
+void segmented_read(std::istream &in, char *buffer, int64_t num_bytes_to_read) {
+    int64_t bytes_remaining = num_bytes_to_read;
+    char *current_ptr = buffer;
+
+    while (bytes_remaining > 0) {
+        // Determine the size of the current chunk to read
+        const auto current_chunk_size = static_cast<std::streamsize>(std::min(bytes_remaining, MAX_SEGMENT_SIZE));
+
+        in.read(current_ptr, current_chunk_size);
+
+        // Check if the read operation failed or didn't read the expected number of bytes
+        if (in.fail() && !in.eof()) { // Failbit set, and not just end of file
```

Comment:
I don't understand why you need to check if in.eof() is false here.  in.fail() should be sufficient to throw an error.  You also have another check after this line, these two checks should be sufficient, I think.

Reply:
Yes, I think this is correct. it doesn't hurt to test eof, but it is redundant.  Changed and pushed.